### DISCUSSION TITLE
Build Error OvmfPkgX64 Unresolved External Symbol

### DISCUSSION THREAD

**Discussion by tawnix** [tianocore/edk2#4780] [Q&A]
Hello, I'm attempting to learn UEFI drivers and this error message has been happening all day. Not sure why an error is happening considering I am literally just trying to build EDK 2 out of the box lol.

I installed NASM, Python3, Windows SDK.

The error message is:
```
HiiDatabase.lib(Image.obj) : error LNK2001: unresolved external symbol memcpy
c:\users\kamer\edk2\Build\OvmfX64\DEBUG_VS2019\X64\MdeModulePkg\Universal\HiiDatabaseDxe\HiiDatabaseDxe\DEBUG\HiiDatabase.dll : fatal error LNK1120: 1 unresolved externals
NMAKE : fatal error U1077: '"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\Hostx86\x64\link.exe" /OUT:c:\users\kamer\edk2\Build\OvmfX64\DEBUG_VS2019\X64\MdeModulePkg\Universal\HiiDatabaseDxe\HiiDatabaseDxe\DEBUG\HiiDatabase.dll /NOLOGO /NODEFAULTLIB /IGNORE:4001 /IGNORE:4281 /OPT:REF /OPT:ICF=10 /MAP /ALIGN:32 /SECTION:.xdata,D /SECTION:.pdata,D /Machine:X64 /LTCG /DLL /ENTRY:_ModuleEntryPoint /SUBSYSTEM:EFI_BOOT_SERVICE_DRIVER /SAFESEH:NO /BASE:0 /DRIVER /DEBUG /WHOLEARCHIVE  @c:\users\kamer\edk2\Build\OvmfX64\DEBUG_VS2019\X64\MdeModulePkg\Universal\HiiDatabaseDxe\HiiDatabaseDxe\OUTPUT\static_library_files.lst' : return code '0x460'
Stop.


build.py...
 : error 7000: Failed to execute command
        C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\Hostx86\x86\nmake.exe /nologo tbuild [c:\users\kamer\edk2\Build\OvmfX64\DEBUG_VS2019\X64\MdeModulePkg\Universal\HiiDatabaseDxe\HiiDatabaseDxe]


build.py...
 : error F002: Failed to build module
        c:\users\kamer\edk2\MdeModulePkg\Universal\HiiDatabaseDxe\HiiDatabaseDxe.inf [X64, VS2019, DEBUG]

```

Not really sure how I'm supposed to fix this issue, im literally just cloning and then building the repo lol.

building via the command "build -t VS2019"

If anyone has information on how to fix this please let me know, thanks!

**Comment by GanbaruTobi** [comment]
I have the same problem.

**Comment by ajfish** [comment]
If you get a random e link failures that usually implies the compiler optimized some random C code into a  memcpy call. This is usually in the form of X = Y and sizeof(Y) > 64-bits. These will tend to be local variables since the compiler can just generate data for globals at compile time. 

The general way to debug this is to` DUMPBIN.EXE <(some flags [1]> Build\OvmfX64\DEBUG_VS2019\X64\MdeModulePkg\Universal\HiiDatabaseDxe\HiiDatabaseDxe\DEBUG\HiiDatabase.dll`
Look at the assembly output to figure out what function the compiler decided to emit the memcpy. Then based on the quality of the disassembly you have figure out the C code, or you go look at the C code in that function and find the bad code and fix it by replacing the = with a CopMem().

[1] I’ve not use VC++ in 10 plus years and don’t have it installed so I don’t remember specifics. 

Thanks,

Andrew Fish

**Comment by microctar** [comment]
> Hello, I'm attempting to learn UEFI drivers and this error message has been happening all day. Not sure why an error is happening considering I am literally just trying to build EDK 2 out of the box lol.
> 
> I installed NASM, Python3, Windows SDK.
> 
> The error message is:
> 
> ```
> HiiDatabase.lib(Image.obj) : error LNK2001: unresolved external symbol memcpy
> c:\users\kamer\edk2\Build\OvmfX64\DEBUG_VS2019\X64\MdeModulePkg\Universal\HiiDatabaseDxe\HiiDatabaseDxe\DEBUG\HiiDatabase.dll : fatal error LNK1120: 1 unresolved externals
> NMAKE : fatal error U1077: '"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\Hostx86\x64\link.exe" /OUT:c:\users\kamer\edk2\Build\OvmfX64\DEBUG_VS2019\X64\MdeModulePkg\Universal\HiiDatabaseDxe\HiiDatabaseDxe\DEBUG\HiiDatabase.dll /NOLOGO /NODEFAULTLIB /IGNORE:4001 /IGNORE:4281 /OPT:REF /OPT:ICF=10 /MAP /ALIGN:32 /SECTION:.xdata,D /SECTION:.pdata,D /Machine:X64 /LTCG /DLL /ENTRY:_ModuleEntryPoint /SUBSYSTEM:EFI_BOOT_SERVICE_DRIVER /SAFESEH:NO /BASE:0 /DRIVER /DEBUG /WHOLEARCHIVE  @c:\users\kamer\edk2\Build\OvmfX64\DEBUG_VS2019\X64\MdeModulePkg\Universal\HiiDatabaseDxe\HiiDatabaseDxe\OUTPUT\static_library_files.lst' : return code '0x460'
> Stop.
> 
> 
> build.py...
>  : error 7000: Failed to execute command
>         C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\Hostx86\x86\nmake.exe /nologo tbuild [c:\users\kamer\edk2\Build\OvmfX64\DEBUG_VS2019\X64\MdeModulePkg\Universal\HiiDatabaseDxe\HiiDatabaseDxe]
> 
> 
> build.py...
>  : error F002: Failed to build module
>         c:\users\kamer\edk2\MdeModulePkg\Universal\HiiDatabaseDxe\HiiDatabaseDxe.inf [X64, VS2019, DEBUG]
> ```
> 
> Not really sure how I'm supposed to fix this issue, im literally just cloning and then building the repo lol.
> 
> building via the command "build -t VS2019"
> 
> If anyone has information on how to fix this please let me know, thanks!

Hello, @tawnix. 

Thanks to @ajfish for the idea, now I've fixed the issue #4917.

**Comment by arinkharkar** [comment]
Run the build command with "-b NOOPT", disabling optimizations so that msvc doesnt call memcpy
